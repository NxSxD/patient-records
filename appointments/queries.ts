import { gql } from "@apollo/react-hooks";
import { Appointment } from "./types";

// Get last 3 appointments
export interface GetAppointmentsVariables {}

export interface GetAppointmentsData {
  appointments: Appointment[]
}

export const GET_APPOINTMENTS = gql`
  query getAppointments {
    appointments(
      limit: 3,
      orderBy: "appointment_time"
    ) {
      id
      doctor
      appointment_time
      location {
        lat
        lng
      }
    }
  }
`;

// Get appointment by ID
export interface GetAppointmentVariables {}

export interface GetAppointmentData {
  appointment: Appointment;
}

export const GET_APPOINTMENT = gql`
  query getAppointment($id: ID!) {
    appointment(id: $id) {
      id
      doctor
      appointment_time
      location {
        lat
        lng
      }
    }
  }
`;

// Add Appointment
export interface AddAppointmentVariables {
  appointment: Omit<Appointment, 'id'>;
}

export interface AddAppointmentData {
  addAppointment: Appointment;
}

export const ADD_APPOINTMENT = gql`
  mutation addAppointment($appointment: NewAppointmentPayload) {
    addAppointment(appointment: $appointment) {
      id
      doctor
      appointment_time
      location {
        lat
        lng
      }
    }
  }
`;