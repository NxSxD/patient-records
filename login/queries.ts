import { gql } from "@apollo/react-hooks";

export interface AuthResponse {
  login: {
    token: string;
  };
}

export interface LoginVariables {
  email: string;
  password: string;
}

export const LOGIN = gql`
  mutation Login($email: String!, $password: String!) {
    login(credentials: {
      email: $email,
      password: $password
    }) {
      token
    }
  }
`;
