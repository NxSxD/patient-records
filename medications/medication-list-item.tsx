import React, { useState, useEffect } from "react";
import styled, { css } from "styled-components";
import { Medication } from "./types";
import {
  ItemInfo,
  InfoLabel,
  ListItemContainer,
  ItemActions,
  ButtonPrimary,
} from "../primitives";
import { SimpleField } from "../components";
import { FormikHelpers, Formik } from "formik";
import { useMutation } from "@apollo/react-hooks";
import {
  UpdateMedicationData,
  UPDATE_MEDICATION,
  UpdateMedicationVariables,
} from "./queries";

interface MedicationListItemProps {
  medication: Medication;
  onItemUpdate: (med: Medication) => void;
}

interface EditValues
  extends Pick<Medication, "name" | "dosage" | "frequency" | "cost"> {}

export const MedicationListItem: React.FC<MedicationListItemProps> = ({
  medication,
  onItemUpdate,
}) => {
  const [updateMedication, { data }] = useMutation<
    UpdateMedicationData,
    UpdateMedicationVariables
  >(UPDATE_MEDICATION);
  const [isEditting, setIsEditting] = useState<boolean>(false);
  const values = {
    name: medication.name,
    dosage: medication.dosage,
    frequency: medication.frequency,
    cost: medication.cost,
  };

  useEffect(() => {
    if (!data) return;

    if (data.updateMedication && data.updateMedication.id) {
      onItemUpdate(data.updateMedication);
    }
  }, [data]);

  const onSubmit = (values: EditValues, helpers: FormikHelpers<EditValues>) => {
    const updatedMed = {
      ...values,
      id: medication.id,
    };

    updateMedication({
      variables: {
        medication: updatedMed,
      },
    });

    setIsEditting(false);
  };

  const editMedication = (_: React.MouseEvent<HTMLButtonElement>) => {
    setIsEditting(true);
  };

  return !isEditting ? (
    <MedicationListContainer>
      <ItemInfo>
        <InfoLabel>Medication</InfoLabel>
        <span>{medication.name}</span>
      </ItemInfo>
      <ItemInfo>
        <InfoLabel>Dosage</InfoLabel>
        <span>{medication.dosage}</span>
      </ItemInfo>
      <ItemInfo>
        <InfoLabel>Frequency</InfoLabel>
        <span>{medication.frequency}</span>
      </ItemInfo>
      <ItemInfo>
        <InfoLabel>Cost</InfoLabel>
        <span>{`$${medication.cost}`}</span>
      </ItemInfo>
      <ItemActions>
        <EditButton onClick={editMedication}>Edit</EditButton>
      </ItemActions>
    </MedicationListContainer>
  ) : (
    <Formik initialValues={values} onSubmit={onSubmit}>
      {({ handleSubmit }) => (
        <form onSubmit={handleSubmit}>
          <MedicationListContainer isEditting={isEditting}>
            <ItemInfo>
              <InfoLabel>Medication</InfoLabel>
              <SimpleField name="name" type="text" />
            </ItemInfo>
            <ItemInfo>
              <InfoLabel>Dosage</InfoLabel>
              <SimpleField name="dosage" type="text" />
            </ItemInfo>
            <ItemInfo>
              <InfoLabel>Frequency</InfoLabel>
              <SimpleField name="frequency" type="text" />
            </ItemInfo>
            <ItemInfo>
              <InfoLabel>Cost</InfoLabel>
              <SimpleField name="cost" type="number" />
            </ItemInfo>
            <ItemActions>
              <EditButton type="submit">Save</EditButton>
            </ItemActions>
          </MedicationListContainer>
        </form>
      )}
    </Formik>
  );
};

const MedicationListContainer = styled(ListItemContainer)<{
  isEditting?: boolean;
}>`
  ${(props) => {
    return (
      props.isEditting &&
      css`
        ${ItemInfo}::not(::last-of-type) {
          padding-right: 1rem;
        }
      `
    );
  }}
`;

const EditButton = styled(ButtonPrimary)`
  min-width: 85px;
  width: 85px;
`;
