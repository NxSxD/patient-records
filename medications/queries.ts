import { gql } from "@apollo/react-hooks";
import { Medication } from "./types";

// Get Medications
export interface GetMedicationsVariables {}

export interface GetMedicationsData {
  medications: Medication[];
}

export const GET_MEDICATIONS = gql`
  query getMedications {
    medications {
      id
      name
      dosage
      frequency
      cost
    }
  }
`;

// Update Medication
export interface UpdateMedicationVariables {
  medication: Medication;
}

export interface UpdateMedicationData {
  updateMedication: Medication;
}

export const UPDATE_MEDICATION = gql`
  mutation updateMed($medication: UpdateMedicationPayload) {
    updateMedication(medication: $medication) {
      id
      name
      dosage
      frequency
    }
  }
`;
