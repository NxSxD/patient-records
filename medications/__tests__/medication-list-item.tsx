import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";

import medicationsJson from "./medications.json";
import { UPDATE_MEDICATION } from "../queries";
import { MedicationListItem } from "../medication-list-item";

const initialMedication = medicationsJson[0];
const updatedMedication = {
  ...initialMedication,
  name: "Medication 1000",
  dosage: "1 Gallon",
  frequency: "1000/always",
  cost: 1000.0,
};

const mocks = [
  {
    request: {
      query: UPDATE_MEDICATION,
      variables: {
        medication: updatedMedication,
      },
    },
    result: {
      data: { updateMedication: updatedMedication },
    },
  },
];

describe("Medication List Item", () => {
  const onUpdate = jest.fn();

  beforeEach(() => {
    onUpdate.mockReset();
  });

  it("should render according to snapshot", () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MedicationListItem
          medication={initialMedication}
          onItemUpdate={onUpdate}
        />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should invoke onUpdate", async () => {
    const { getByText, getByLabelText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MedicationListItem
          medication={initialMedication}
          onItemUpdate={onUpdate}
        />
      </MockedProvider>
    );

    fireEvent.click(getByText("Edit"));

    fireEvent.input(getByLabelText(/Name-input/), {
      target: { value: updatedMedication.name },
    });

    fireEvent.change(getByLabelText(/Dosage-input/), {
      target: { value: updatedMedication.dosage },
    });

    fireEvent.input(getByLabelText(/Frequency-input/), {
      target: { value: updatedMedication.frequency },
    });

    fireEvent.input(getByLabelText(/Cost-input/), {
      target: { value: updatedMedication.cost },
    });

    fireEvent.click(getByText("Save"));

    await waitFor(() => {
      expect(getByText("Save")).toBeInTheDocument();
    });

    expect(onUpdate).toHaveBeenCalled();
  });
});
