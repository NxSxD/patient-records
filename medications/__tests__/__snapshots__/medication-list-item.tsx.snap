// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Medication List Item should render according to snapshot 1`] = `
.c2 {
  font-size: 0.9rem;
  color: var(--text-label-color);
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c0 {
  height: 65px;
  border-radius: 4px;
  border: 1px solid #ccc;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 0 1rem;
  margin-bottom: 1rem;
}

.c0:hover {
  cursor: pointer;
}

.c4 {
  height: 45px;
  padding: 0 1rem;
  font-size: 1.2rem;
  border: none;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: background-color 250ms;
  transition: background-color 250ms;
  min-width: 120px;
  background-color: var(--primary-color);
  color: #fff;
  min-width: 85px;
  width: 85px;
}

.c4:hover {
  cursor: pointer;
  background-color: var(--primary-alt-color);
}

.c4:focus {
  outline: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Medication
      </span>
      <span>
        Medication 1
      </span>
    </div>
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Dosage
      </span>
      <span>
        1 Tablet
      </span>
    </div>
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Frequency
      </span>
      <span>
        1/day
      </span>
    </div>
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Cost
      </span>
      <span>
        $5
      </span>
    </div>
    <div
      class="c3"
    >
      <button
        class="c4"
      >
        Edit
      </button>
    </div>
  </div>
</div>
`;
